src/Interpreters/ProcessList.cpp

```diff
@@ -430,23 +433,62 @@ QueryStatus * ProcessList::tryGetProcessListElement(const String & current_query
 
 CancellationCode ProcessList::sendCancelToQuery(const String & current_query_id, const String & current_user, bool kill)
 {
-    auto lock = safeLock();
+    QueryStatus * elem;
+
+    /// Cancelling the query should be done without the lock.
+    ///
+    /// Since it may be not that trivial, for example in case of distributed
+    /// queries it tries to cancel the query gracefully on shards and this can
+    /// take a while, so acquiring a lock during this time will lead to wait
+    /// all new queries for this cancellation.
+    ///
+    /// Another problem is that it can lead to a deadlock, because of
+    /// OvercommitTracker.
+    ///
+    /// So here we first set is_cancelling, and later reset it.
+    /// The ProcessListEntry cannot be destroy if is_cancelling is true.
+    {
+        auto lock = safeLock();
+        elem = tryGetProcessListElement(current_query_id, current_user);
+        if (!elem)
+            return CancellationCode::NotFound;
+        elem->is_cancelling = true;
+    }
 
-    QueryStatus * elem = tryGetProcessListElement(current_query_id, current_user);
+    SCOPE_EXIT({
+        DENY_ALLOCATIONS_IN_SCOPE;
 
-    if (!elem)
-        return CancellationCode::NotFound;
+        auto lock = unsafeLock();
+        elem->is_cancelling = true;
+        cancelled_cv.notify_all();
+    });
 
     return elem->cancelQuery(kill);
 }
 
 
 void ProcessList::killAllQueries()
 {
-    auto lock = safeLock();
+    std::list<QueryStatus *> cancelled_processes;
+
+    SCOPE_EXIT({
+        auto lock = safeLock();
+        for (auto & cancelled_process : cancelled_processes)
+            cancelled_process->is_cancelling = false;
```

Comment:
Don't we need to do a `notify_all` here?

Reply:
Indeed, nice catch! Fixed.